### PR TITLE
VZ-7693 receive alerts worker

### PR DESCRIPTION
**Pull request by mbenmclean**
This adds a receive alerts worker that perfoms an update on the VZ CR and configures Verrazzano to send alerts as POSTS. It also receives the alert POSTs and generates events from them.

### CODE REVIEW

File: tools/psr/backend/workers/http/receiveAlerts/receiveAlerts.go [hunk 1]
```diff
@@ -0,0 +1,230 @@
+// Copyright (c) 2022, Oracle and/or its affiliates.
```
Comment (pfmackin) [inline tools/psr/backend/workers/http/receiveAlerts/receiveAlerts.go:1]: This needs to go into workers/prometheus.  The HTTP part is just an implementation detail.  I am assuming all other components can use this code to get alerts.

File: tools/psr/backend/workers/http/receiveAlerts/receiveAlerts.go [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+	"github.com/verrazzano/verrazzano/tools/psr/backend/spi"
+
+	"github.com/prometheus/client_golang/prometheus"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// metricsPrefix is the prefix that is automatically pre-pended to all metrics exported by this worker.
+const metricsPrefix = "http_alerts"
```
Comment (pfmackin) [inline tools/psr/backend/workers/http/receiveAlerts/receiveAlerts.go:31]: should be prom_alert or prometheus_alert

File: tools/psr/backend/workers/http/receiveAlerts/receiveAlerts.go [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+		}
+		if alertStatus == "firing" {
+			atomic.AddInt64(&w.workerMetrics.alertsFiringCount.Val, 1)
+		} else if alertStatus == "resolved" {
+			atomic.AddInt64(&w.workerMetrics.alertsResolvedCount.Val, 1)
+		} else {
+			log.Errorf("alert received with unknown status: %s", alertStatus)
+		}
+
+		event := corev1.Event{
```
Comment (pfmackin) [inline tools/psr/backend/workers/http/receiveAlerts/receiveAlerts.go:146]: i would prefer if this was in backend/pkg/k8s/event.go

File: tools/psr/backend/workers/http/receiveAlerts/receiveAlerts.go [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+	if err != nil {
+		return err
+	}
+	var m psrprom.AlertmanagerConfigModifier
+	m.ModifyCR(cr)
+
+	return psrvz.UpdateVerrazzano(c.VzInstall, cr)
+}
+
+func createAlertmanagerOverridesCM(log vzlog.VerrazzanoLogger, conf config.CommonConfig) error {
```
Comment (pfmackin) [inline tools/psr/backend/workers/http/receiveAlerts/receiveAlerts.go:188]: Similar comment, can this be moved to pkg/prometheus/alert.go

File: tools/psr/backend/workers/http/receiveAlerts/receiveAlerts_test.go [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+					found++
+				}
+			}
+			assert.Equal(t, 1, found)
+		})
+	}
+}
+
+//
+//// TestDoWork tests the DoWork method
```
Comment (pfmackin) [inline tools/psr/backend/workers/http/receiveAlerts/receiveAlerts_test.go:107]: remove commented-out code

**Comment by pfmackin** [comment]
Are you planning on adding rules YAML?

File: tools/psr/backend/workers/http/receiveAlerts/receiveAlerts.go (hunk 1 above)
Comment (mbenmclean) [reply tools/psr/backend/workers/http/receiveAlerts/receiveAlerts.go:1]: Ok I'll refactor it into prometheus

**Comment by mbenmclean** [comment]
> Are you planning on adding rules YAML?

I was thinking of doing that as a separate PR because it's kind distinct from this change

File: tools/psr/backend/workers/http/receiveAlerts/receiveAlerts.go (hunk 3 above)
Comment (mbenmclean) [reply tools/psr/backend/workers/http/receiveAlerts/receiveAlerts.go:146]: We don't really use events outside of alertmanager so I think it makes sense to leave in place for now. This code is pretty specialized don't think it makes sense to break it out.

File: tools/psr/backend/workers/http/receiveAlerts/receiveAlerts.go (hunk 4 above)
Comment (mbenmclean) [reply tools/psr/backend/workers/http/receiveAlerts/receiveAlerts.go:188]: moved

File: tools/psr/backend/workers/promtheus/alerts/alerts.go [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+
+		event := corev1.Event{
+			ObjectMeta: v1.ObjectMeta{
+				Name:      "psr-alert-" + alertName,
+				Namespace: config.PsrEnv.GetEnv(config.PsrWorkerNamespace),
+			},
+			InvolvedObject: corev1.ObjectReference{
+				Namespace: config.PsrEnv.GetEnv(config.PsrWorkerNamespace),
+			},
+			Type: "Alert",
```
Comment (michael-cico) [inline tools/psr/backend/workers/promtheus/alerts/alerts.go:153]: Does the Type have to be something that doesn't conflict with other types?  i.e., does this type have to be qualified?

File: tools/psr/manifests/charts/worker/templates/deployment.yaml [hunk 7]
```diff
@@ -32,12 +32,12 @@ spec:
...
             - name: NAMESPACE
-              valueFrom:
-                fieldRef:
-                  fieldPath: metadata.namespace
+              value: {{ .Release.Namespace }}
+            - name: RELEASE_NAME
+              value: {{ .Release.Name }}
           ports:
-            - containerPort: 8080
+            - containerPort: 9090
```
Comment (michael-cico) [inline tools/psr/manifests/charts/worker/templates/deployment.yaml:41]: Just curious, what triggered the port change?

File: tools/psr/backend/workers/promtheus/alerts/alerts.go [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+	"github.com/verrazzano/verrazzano/tools/psr/backend/spi"
+
+	"github.com/prometheus/client_golang/prometheus"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// metricsPrefix is the prefix that is automatically pre-pended to all metrics exported by this worker.
+const metricsPrefix = "prom_alerts"
```
Comment (pfmackin) [inline tools/psr/backend/workers/promtheus/alerts/alerts.go:31]: please change to prom_alert

File: tools/psr/backend/workers/promtheus/alerts/alerts.go (hunk 6 above)
Comment (mbenmclean) [reply tools/psr/backend/workers/promtheus/alerts/alerts.go:153]: Not that I know of I look online

**Review by pfmackin — APPROVED** [review]

**Pull request merged**